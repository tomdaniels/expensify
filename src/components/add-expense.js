import React from 'react';
import ExpenseForm from './expense-form';

const AddExpensePage = () => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm
            onSubmit={(expense) => {
                console.log(expense);
            }}
        />
    </div>
);

export default AddExpensePage;