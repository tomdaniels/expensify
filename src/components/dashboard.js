import React from 'react';
import ExpenseList from './expense-list';
import ExpenseSummary from './expense-summary';
import ExpenseListFilters from './expense-list-filters';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseSummary />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;