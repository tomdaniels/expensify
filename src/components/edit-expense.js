import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import ExpenseForm from './expense-form';
import { startEditExpenses, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    state = {
      removeClick: false
    };
    onSubmit = (expense) => {
        this.props.startEditExpenses(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
        this.setState((prevState) => ({
            removeClick: !prevState.removeClick,
        }));
    };

    handleRemoveConfirm = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push('/');
    };

    handleRemoveCancel = () => {
        this.setState((prevState) => ({
            removeClick: !prevState.removeClick,
        }));
    };

    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page_header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm
                        expense={this.props.expense}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Expense</button>
                    <Modal
                        isOpen={this.state.removeClick}
                        onRequestClose={this.handleRemoveCancel}
                        contentLabel="Confirm remove"
                        closeTimeoutMS={200}
                        className="modal"
                        ariaHideApp={false}
                    >
                        <p className="modal__title">Are you sure you want to remove this expense?</p>
                        <div className="modal__button-wrap">
                            <button className="button button--link" onClick={this.handleRemoveConfirm}>Yes</button>
                            <button className="button button--link" onClick={this.handleRemoveCancel}>No</button>
                        </div>
                    </Modal>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) =>
        expense.id === props.match.params.id
    )
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpenses: (id, expense) => dispatch(startEditExpenses(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);