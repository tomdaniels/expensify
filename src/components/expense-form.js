import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            isRecurring: props.expense ? props.expense.isRecurring : false,
            calendarFocused: false,
            errorState: ''
        }
    }

    onDescriptionChange = (e) => {
      const description = e.target.value;
      this.setState(() => ({
          description
      }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({
            note
        }));
    };

    onAmountChange = (e) => {
      const amount = e.target.value;

      if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
          this.setState(() => ({
              amount
          }));
      }
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({
                createdAt
            }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused
        }));
    };

    toggleRecurring = () => {
        this.setState((prevState) => ({
            isRecurring: !prevState.isRecurring,
        }));
    };

    onSubmit = (event) => {
        event.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                errorState: 'Please provide description and amount.'
            }));
        } else {
            this.setState(() => ({
                errorState: ''
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
                isRecurring: this.state.isRecurring,
            });
        }
    };

    render() {
        return (
          <form className="form" onSubmit={this.onSubmit}>
              {
                  this.state.errorState &&
                  <p className="form__error">{this.state.errorState}</p>
              }
              <input
                  type="text"
                  placeholder="Description"
                  className="text-input"
                  autoFocus
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
              />
              <input
                  value={this.state.amount}
                  className="text-input"
                  type="text"
                  placeholder="Amount"
                  onChange={this.onAmountChange}
              />
              <input
                  type="checkbox"
                  selected="selected"
                  onClick={this.toggleRecurring}
              />
              <SingleDatePicker
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calendarFocused}
                  onFocusChange={this.onFocusChange}
                  numberOfMonths={1}
                  isOutsideRange={() => false}
                  block
              />
              <textarea
                  className="text-area"
                  value={this.state.note}
                  onChange={this.onNoteChange}
                  placeholder="Add a note for your expense (optional)"
              >
              </textarea>
              <div>
                  <button className="button">
                      {
                          this.props.expense ? 'Save Expense' : 'Add Expense'
                      }
                  </button>
              </div>
          </form>
        );
    }
}

export default ExpenseForm;