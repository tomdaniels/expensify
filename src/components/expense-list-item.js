import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import Icon from 'react-icons/lib/fa/calendar';

const ExpenseListItem = ({ id, description, amount, createdAt, isRecurring }) => (
    <Link
        className="list-item"
        to={`/edit/${id}`}
    >
        <div>
            <h3 className="list-item__title">{description}</h3>
            {
                isRecurring &&
                        <Icon
                            className="list-item__recurring-icon"
                            data-tip="test"
                        />
            }
            <span className="list-item__subtitle">{moment(createdAt).format('Do MMMM, YYYY')}</span>
            {
                isRecurring &&
                <p className="list-item__recurring">
                            recurring expense
                        </p>
            }
        </div>
        <div>
            <h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
        </div>
    </Link>
);



export default ExpenseListItem;