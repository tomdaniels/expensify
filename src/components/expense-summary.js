import React from 'react';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpenseTotal from '../selectors/expenses-total';
import hiddenExpensesCount from '../selectors/hidden-expenses'

export const ExpenseSummary = ({ expenseCount, total, hidden }) => {
    const expenseWord = expenseCount.length === 1 ? 'expense' : 'expenses';
    const hiddenWord = hidden === 1 ? 'expense' : 'expenses';
    const formattedExpensesTotal = numeral(total / 100).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page_header__title">Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span></h1>
                <h4 className="page_header__title subtitle"><span>{hidden}</span> hidden {hiddenWord}</h4>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expenses</Link>
                </div>
            </div>
        </div>
    )
};

const mapStateToProps = state => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        total: selectExpenseTotal(visibleExpenses),
        hidden: hiddenExpensesCount(state.expenses, visibleExpenses)
    }
};

export default connect(mapStateToProps)(ExpenseSummary);