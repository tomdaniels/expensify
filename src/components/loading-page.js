import React from 'react';

const LoadingPage = () => (
    <div className="loader">
        <img className="loader__image" src="/public/images/loader.gif"/>
    </div>
);

export default LoadingPage;