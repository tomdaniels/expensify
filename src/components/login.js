import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ({ startLogin }) => (
    <div className="login-page__box-layout">
        <div className="login-page__box">
            <h1 className="login-page__title">Expensify</h1>
            <p>It's time to get your expenses under control</p>
            <button onClick={startLogin}>Login</button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);