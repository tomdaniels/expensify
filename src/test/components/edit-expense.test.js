import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/edit-expense';
import expenses from '../fixtures/expenses';

let startEditExpenses, startRemoveExpense, history, wrapper;

beforeEach(() => {
    startEditExpenses = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage
            startEditExpenses={startEditExpenses}
            startRemoveExpense={startRemoveExpense}
            history={history}
            expense={expenses[2]}
        />
    );
});

test('should render the edit expense page correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpenses).toHaveBeenLastCalledWith(expenses[2].id, expenses[2]);
});

// test('should handle startRemoveExpense', () => {
//     wrapper.find('button').at(0).simulate('click');
//     expect(wrapper.find('Modal'));
//     expect(startRemoveExpense).toHaveBeenLastCalledWith({
//         id: expenses[2].id
//     });
// });