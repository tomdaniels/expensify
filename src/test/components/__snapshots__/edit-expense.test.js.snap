// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the edit expense page correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page_header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "water",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
    <Modal
      ariaHideApp={false}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={200}
      contentLabel="Confirm remove"
      isOpen={false}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <p
        className="modal__title"
      >
        Are you sure you want to remove this expense?
      </p>
      <div
        className="modal__button-wrap"
      >
        <button
          className="button button--link"
          onClick={[Function]}
        >
          Yes
        </button>
        <button
          className="button button--link"
          onClick={[Function]}
        >
          No
        </button>
      </div>
    </Modal>
  </div>
</div>
`;
